File: service-request-frontend/src/services/categoryService.js
import api from '../api/axios';

/**
 * Category Service
 * Handles API calls for service categories and request types
 * Used by Create Request flow
 */
const categoryService = {
    /**
     * Get all active service categories
     * GET /api/service-categories
     */
    getCategories: async () => {
        console.log('📡 API Call: GET /api/service-categories');
        try {
            const response = await api.get('/service-categories');
            console.log('✅ Categories loaded:', response.data.length, 'categories');
            console.log('Categories:', response.data);
            return response.data;
        } catch (error) {
            console.error('❌ Error loading categories:', error);
            console.error('Error details:', error.response?.data);
            throw error;
        }
    },

    /**
     * Get category by ID
     * GET /api/service-categories/{id}
     */
    getCategoryById: async (categoryId) => {
        console.log(`📡 API Call: GET /api/service-categories/${categoryId}`);
        try {
            const response = await api.get(`/service-categories/${categoryId}`);
            console.log('✅ Category loaded:', response.data);
            return response.data;
        } catch (error) {
            console.error(`❌ Error loading category ${categoryId}:`, error);
            throw error;
        }
    },

    /**
     * Get request types for a category
     * GET /api/service-categories/{categoryId}/types
     */
    getCategoryTypes: async (categoryId) => {
        console.log(`📡 API Call: GET /api/service-categories/${categoryId}/types`);
        try {
            const response = await api.get(`/service-categories/${categoryId}/types`);
            console.log('✅ Request types loaded:', response.data.length, 'types');
            console.log('Request types:', response.data);
            return response.data;
        } catch (error) {
            console.error(`❌ Error loading request types for category ${categoryId}:`, error);
            console.error('Error details:', error.response?.data);
            throw error;
        }
    },

    /**
     * Get all active request types
     * GET /api/service-categories/types/all
     */
    getAllTypes: async () => {
        console.log('📡 API Call: GET /api/service-categories/types/all');
        try {
            const response = await api.get('/service-categories/types/all');
            console.log('✅ All request types loaded:', response.data.length, 'types');
            return response.data;
        } catch (error) {
            console.error('❌ Error loading all request types:', error);
            throw error;
        }
    }
};

export default categoryService;

File: service-request-backend/src/main/java/com/servicedesk/service/CategoryService.java
package com.servicedesk.service;

import com.servicedesk.entity.RequestType;
import com.servicedesk.entity.ServiceCategory;
import com.servicedesk.repository.RequestTypeRepository;
import com.servicedesk.repository.ServiceCategoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Service for managing request categories and types
 */
@Service
@Transactional
public class CategoryService {

    @Autowired
    private ServiceCategoryRepository categoryRepository;

    @Autowired
    private RequestTypeRepository requestTypeRepository;

    /**
     * Get all active categories
     */
    public List<ServiceCategory> getAllActiveCategories() {
        return categoryRepository.findByIsActiveTrueOrderByNameAsc();
    }

    /**
     * Get category by ID
     */
    public ServiceCategory getCategoryById(Long id) {
        return categoryRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Category not found with id: " + id));
    }

    /**
     * Get all request types for a category
     */
    public List<RequestType> getTypesByCategory(Long categoryId) {
        System.out.println("=== FETCHING REQUEST TYPES ===");
        System.out.println("Category ID: " + categoryId);

        try {
            List<RequestType> types = requestTypeRepository.findByCategoryIdAndIsActiveTrue(categoryId);
            System.out.println("Found " + types.size() + " types for category " + categoryId);
            return types;
        } catch (Exception e) {
            System.err.println("Error fetching types for category " + categoryId + ": " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("Failed to fetch request types", e);
        }
    }

    /**
     * Get request type by ID
     */
    public RequestType getTypeById(Long id) {
        return requestTypeRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Request type not found with id: " + id));
    }

    /**
     * Get all active request types
     */
    public List<RequestType> getAllActiveTypes() {
        return requestTypeRepository.findByIsActiveTrue();
    }
}
